src/KioskMode/README.md

```diff
@@ -0,0 +1,102 @@
+  ILIAS is used as a Tool-Provider.
+* The Learning Sequence requires a kiosk mode to maintain control over the general
+  display and navigation.
+
+This library facilitates the implementation of objects that can be viewed in a kiosk
+mode and of features that display these objects in a kiosk mode. The provided means
+aim to make it possible for objects to implement a kiosk mode only once and then be
+used by all features that need to display objects in that kiosk mode. On the other
+hand, features that require viewing other objects in a kiosk mode should be able to
+use all objects that implement a kiosk mode.
+
+This library does neither implement the kiosk-mode for certain objects, nor implements
+a view on these objects. It rather provides interfaces that describe how consumers
+and providers of a kiosk mode should collaborate and functionality that helps to
+implement both sides of the collaboration.
+
+This library is complemented by the [Kiosk-Mode-Service](../../Services/KioskMode/README.md)
+that provides functionality to construct views for certain objects and concrete
+database functionality.
+
+## Architecture
+
+A **provider** of a kiosk mode is an ILIAS-object that may be viewed in in a kiosk
+mode. A **player** is any feature of ILIAS that wants to display ILIAS-objects in
+the reduced way defined by the kiosk mode.
+
+The **player** may request a **view** for a certain **provider** at the kiosk mode
+service. The **view** then is the exclusive way for the **player** and the
+**provider** to communicate.
+
+The **view** implements the following functionality to display and control the
+providing object via the kiosk mode:
+
+* It provides **controls** to the player that can be used to modify the state of
+  the view.
+* It provides a function that **updates** a **state** based on the use of **controls**
+  to track the condition of the view.
+* It provides a function to **render** the **state** to allow the player to display
+  the object in a kiosk mode.
+
+The **player** in turn uses that functions of the view to display and execute the
+kiosk mode of the object:
+
+* It asks the view for **controls** and displays them in an appropriate way.
+* It processes user input for the **controls** and asks the view to **update** the
+  **state**. It makes sure that the state is persisted appropriately.
+* It asks the view to **render** the **state** and displays the result in an
+  appropritate way and visual environment, most possibly in conjunction with the
+  controls.
+
+
+## Implementing a Provider
+
+Please use `ilKioskModeView` in `Services\KioskMode\classes\` as a base to implement
+this for ILIAS-objects.
+
+* `State` MUST only contain information about the view (not about LP, completion, ...)
+* `View` MUST NOT maintain state of the view internally but put it into `State`.
+* `View::render` MUST NOT render controls that trigger a page change. It MAY provide
```

Comment:
Jour Fixe, 18 JUL 2018 : Please change the rule as follows: "`View::render` SHOULD NOT render controls that trigger a page change. It MAY provide …" to allow internal linking in components.

Reply:
My current understanding the meaning of this is:
If the view renders controls that trigger a page change the view MUST open in a new browser window. And this browser window will **not provide the** Kiosk mode anymore. Is this how it is meant?

This does not become clear.  This means all interactivity within the object MUST be provided through kiosk mode controls, right? This is what leads to the high costs if we want to keep the existing LTI feature set here, imo. We need to rewrite all interactive views.

Reply:
Yes, the idea is, that controls need to be provided via kiosk mode. Also: Nobody demanded that LTI need to be implemented via this mechanic here.

With @eqsoft I discussed this proposal and its relation to LTI and we both agreed that the current proposal is to weak to implement LTI functionality. But: the model I propose here, which basically comes down to moving general rendering and control of user interactions out of the view itself to another component, seemed suitable for the LTI-mode as well to both of us. The controls that could be provided would need to be richer for that case.

Reply:
"I propose here, which basically comes down to moving general rendering and control of user interactions out of the view itself to another component."

But why would this be now inside the responsibility of a a kiosk mode player component? This would be a very fundamental move affecting the whole platform so the "other component" should be a much more core UI control component imo.

Reply:
> "Nobody demanded that LTI need to be implemented via this mechanic here."

But the proposal is about a "General Kiosk Mode" and LTI is named as one of the three main scenarios in the introduction.